src/app/cluster/duck/sagas.ts

```diff
@@ -65,15 +65,91 @@ function* addClusterRequest(action) {
   );
   const migClusterResource = new MigResource(MigResourceKind.MigCluster, migMeta.namespace);
 
+  // Ensure that none of objects that make up a cluster already exist
   try {
-    const clusterAddResults = yield Promise.all([
+    const getResults = yield Q.allSettled([
+      client.get(clusterRegResource, clusterReg.metadata.name),
+      client.get(secretResource, tokenSecret.metadata.name),
+      client.get(migClusterResource, migCluster.metadata.name),
+    ]);
+
+    const alreadyExists = getResults.reduce((exists, res) => {
+      return res.value && res.value.status === 200 ?
+        [...exists, { kind: res.value.data.kind, name: res.value.data.metadata.name }] :
+        exists;
+    }, []);
+
+    if(alreadyExists.length > 0) {
+      throw new Error(alreadyExists.reduce((msg, v) => {
+        return msg + `- kind: "${v.kind}", name: "${v.name}"`;
+      }, 'Some cluster objects already exist '));
+    }
+  } catch(err) {
+    console.error(err.message);
+    yield put(ClusterActions.setClusterAddEditStatus(createAddEditStatusWithMeta(
+      AddEditState.Critical,
+      AddEditMode.Add,
+      err.message,
+      '',
+    )));
+    return;
+  }
+
+  // None of the objects already exist, so try to create all of the objects
+  // If any of the objects actually fail creation, we need to rollback the others
+  // so the clusters are created, or fail atomically
+  try {
+    const clusterAddResults = yield Q.allSettled([
       client.create(clusterRegResource, clusterReg),
       client.create(secretResource, tokenSecret),
       client.create(migClusterResource, migCluster),
     ]);
 
+    // If any of the attempted object creation promises have failed, we need to
+    // rollback those that succeeded so we don't have a halfway created "Cluster"
+    // A rollback is only required if some objects have actually *succeeded*,
+    // as well as failed.
+    const isRollbackRequired =
+      clusterAddResults.find(res => res.state === 'rejected') &&
+      clusterAddResults.find(res => res.state === 'fulfilled');
+
+    if(isRollbackRequired) {
+      const kindToResourceMap = {
+        Cluster: clusterRegResource,
+        MigCluster: migClusterResource,
+        Secret: secretResource,
+      };
+
+      // The objects that need to be rolled back are those that were fulfilled
+      const rollbackObjs = clusterAddResults.reduce((rollbackAccum, res) => {
+        return res.state === 'fulfilled' ?
+          [...rollbackAccum, { kind: res.value.data.kind, name: res.value.data.metadata.name }] :
+          rollbackAccum;
+      }, []);
+
+      const rollbackResults = yield Q.allSettled(rollbackObjs.map(r => {
```

Comment:
👍

Reply:
Only 2 comments with this:

- Check connection button shows ready after clicking 2 times. Any changes we can make this a better experience?

- Are we changing the check connection button to check with the new values entered in the form? Or is the current behavior ok?